### ISSUE TITLE
How can I turn output and loss into an array?

### ISSUE THREAD

**Issue by NoteDancing** [google/jax#15136]
### Description

[here](https://colab.research.google.com/drive/1M2fO8yhlTVVLkdMT8c9Pqsgs92mlP1VS?usp=sharing)

### What jax/jaxlib version are you using?

jax v0.4.6

### Which accelerator(s) are you using?

GPU

**Comment by jakevdp** [comment]
Your code is failing because you are JIT-compiling an impure function (see [JAX Sharp Bits: Pure Functions](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#pure-functions) for a discussion of this)

The problem is in this line:
```python
  @functools.partial(jit, static_argnums=(0,))
  def opt(self):
    grads=grad(self.nn.loss)(self.nn.param,self.data,self.labels)
    self.param=[(w-self.nn.lr*dw,b-self.nn.lr*db) for (w,b),(dw,db) in zip(self.nn.param,grads)]
    return
```
Your options are either (1) don't jit-compile this function, or (2) don't modify `self` within this jit-compiled function.

Another issue here is that you are marking `self` as static, when it is not static (you are mutating it within the function and elsewhere). For a discussion of how to use JIT properly with class methods, see [FAQ: how to use JIT with methods](https://jax.readthedocs.io/en/latest/faq.html#how-to-use-jit-with-methods).

**Comment by jakevdp** [comment]
I'm going to close the issue because I think this answers the question - if not feel free to respond here!

**Comment by NoteDancing** [comment]
@jakevdp Thank you for your answer.
